cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_fp_impl.hpp

```diff
@@ -154,66 +154,58 @@ sycl::event kernels_fp<Float>::select(sycl::queue& queue,
-                    if (global_row_id >= last_row)
-                        return;
-                    auto sg = item.get_sub_group();
-                    const std::int64_t sg_id = sg.get_group_id()[0];
-                    const std::int64_t wg_id = item.get_global_id(1);
-                    const std::int64_t sg_num = sg.get_group_range()[0];
-                    const std::int64_t sg_global_id = wg_id * sg_num + sg_id;
-                    if (sg_global_id >= row_count)
-                        return;
-                    const std::int64_t in_offset = sg_global_id * stride;
-                    const std::int64_t out_offset = sg_global_id;
-
-                    const std::int64_t local_id = sg.get_local_id()[0];
-                    const std::int64_t local_range = sg.get_local_range()[0];
-
-                    std::int32_t index = -1;
-                    Float value = fp_max;
-                    for (std::int64_t i = local_id; i < cluster_count_as_int32; i += local_range) {
-                        const Float cur_val =
-                            distances_ptr[in_offset + i] + centroid_squares_ptr[i];
-                        if (cur_val < value) {
-                            index = static_cast<std::int32_t>(i);
-                            value = cur_val;
-                        }
-                    }
 
-                    sg.barrier();
-
-                    const Float final_value =
-                        sycl::reduce_over_group(sg, value, sycl::ext::oneapi::minimum<Float>());
-                    const bool present = (final_value == value);
-                    const std::int32_t pos =
-                        sycl::exclusive_scan_over_group(sg,
-                                                        present ? 1 : 0,
-                                                        sycl::ext::oneapi::plus<std::int32_t>());
-                    const bool owner = present && pos == 0;
-                    const std::int32_t final_index =
-                        -sycl::reduce_over_group(sg,
-                                                 owner ? -index : 1,
-                                                 sycl::ext::oneapi::minimum<std::int32_t>());
+            cgh.parallel_for<select_min_distance<Float>>(range, [=](sycl::nd_item<2> item) {
+                constexpr sycl::ext::oneapi::minimum<Float> minimum_val;
+                constexpr sycl::ext::oneapi::minimum<std::int64_t> minimum_idx;
 
-                    if (local_id == 0) {
-                        indices_ptr[out_offset] = final_index;
-                        selection_ptr[out_offset] = final_value;
-                    }
-                });
+                const std::int64_t row = item.get_global_id(0) + first_row;
+
+                if (last_row <= row)
+                    return;
+                auto sg = item.get_sub_group();
+                std::int64_t local_id = item.get_local_id(1);
```

Comment:
Please use `std::int32_t`. This value is limited by number of Work Items in the Sub Group which is less or equal to the number of Work Items in Work Group. Thus it is a pretty small value by definition.

Reply:
```suggestion
                std::int32_t local_id = item.get_local_id(1);
```